### ISSUE TITLE
tab complete results in duplicate entries

### ISSUE THREAD

**Issue by wcbeard** [ipython/ipython#10960]
Here's a screenshot of completing for `a = np.ndarray(3)`

![screen shot 2017-12-21 at 10 39 59 am](https://user-images.githubusercontent.com/379413/34262981-f13412c6-e63b-11e7-8858-34b314e5369b.png)

They don't seem to be duplicated in the __dir__ method:

```
Counter(a.__dir__()).most_common(4)
=> [('__repr__', 1), ('__hash__', 1), ('__str__', 1), ('__lt__', 1)]
```

Sys info
```
$ python -c "import IPython; print(IPython.sys_info())"
{'commit_hash': 'ca5443062',
 'commit_source': 'installation',
 'default_encoding': 'UTF-8',
 'ipython_path': '/Users/wbeard/miniconda3/envs/test/lib/python3.6/site-packages/IPython',
 'ipython_version': '6.2.1',
 'os_name': 'posix',
 'platform': 'Darwin-16.7.0-x86_64-i386-64bit',
 'sys_executable': '/Users/wbeard/miniconda3/envs/test/bin/python',
 'sys_platform': 'darwin',
 'sys_version': '3.6.3 | packaged by conda-forge | (default, Dec  9 2017, '
                '16:20:51) \n'
                '[GCC 4.2.1 Compatible Apple LLVM 6.1.0 (clang-602.0.53)]'}
```

Conda install file

```
$ cat environment.yml | egrep -v "(^ *#.*|^$)"
name: test
dependencies:
- python=3
- flake8
- pandas
- jupyter
- notebook
- pip:
  - line-profiler
  - jupyter_contrib_nbextensions
  - python-dotenv>=0.5.1
```

**Comment by wcbeard** [comment]
Actually, this is only a problem in the jupyter notebook, not the ipython interpreter. I can close here and open there if someone can confirm that's what I should do.

**Comment by takluyver** [comment]
Other people already have - https://github.com/jupyter/notebook/issues/3145 - but I think the issue is actually coming from IPython (it's just not visible in the terminal interface).

**Comment by takluyver** [comment]
@Carreau I think I've tracked this down. The completions are deduplicated before going through `rectify_completions`. At this point (completing `a.<tab>`), one system gives `a.foo`, and the other just gives `foo`, so they're not duplicates. When they're rectified, `foo` becomes `a.foo`, but the deduplication has already happened.

Presumably we want to rearrange things so they are rectified and then deduplicated. Have you got time to work on this? You're more familiar with that bit of the code than I am.

**Comment by Carreau** [comment]
Yeah, there is already a function that deduplicate not based only on the completions themselves, but also based on the completions once applied.  I hadn't  had a chance to send a patch.

There is a `_deduplicate_completions` already available  in `IPython.core.completer` and used only by prompt_toolkit.

It's probably just a matter of plugging it in the right place in ipykernel.

A more permanent fix would be to use the new completion API and give feedback on it with notebooks fromtend.

**Comment by takluyver** [comment]
I think we should make it so you get the Jedi completions by default and *not* our Python completions, and have a config option which simultaneously turns jedi integration off and our own completions back on. (Our magic and filename completions would be used either way)

If we combine them, as we currently do, we're not going to get reports of things missing from the Jedi completions, because they're filled in by IPython's completions.

**Comment by paulochf** [comment]
I've just got onto this issue.

Problem comes from ipykernel==4.7.0. I did `pip install ipykernel==4.6.1` and it got fixed.

**Comment by takluyver** [comment]
I think this should be fixed in master by #10969.

**Comment by takluyver** [comment]
Closing as probably fixed. I'm trying to work towards a new release soon.
